src/Core/Services/MpambaService.cs

```diff
@@ -17,34 +17,40 @@ public MpambaService(string message, string phoneNumber)
 
         public Payment GeneratePayment()
         {
-            var amountRegex = new Regex("((?<=(Amount: )|(Amt: ))(.*?)(?=M))|((?<=(recieved ))(.*?)(?=M))");
-            var referenceRegex = new Regex(@"(?<=Ref: )(.*?)(?=\s)");
-            var bankNameRegex = new Regex("(?<=from )(.*?)(?= on)");
-            var fromAgentRegex = new Regex("(Cash In)");
+            var amountRegex = new Regex("((?<=(Amount:)|(Amt:))(.*?)(?=M))|((?<=(recieved))(.*?)(?=M))");
             var amount = Decimal.Parse(amountRegex.Match(_message).ToString().Trim());
-            var reference = referenceRegex.Match(_message).ToString();
-            var fromAgent = fromAgentRegex.IsMatch(_message);
-            var BankName = bankNameRegex.Match(_message).ToString();
+            var reference = CreateReference();
+            var agentName = Regex.IsMatch(_message,"(CashIn)") ? GetSenderName() : "Missing";
+            var SenderName = Regex.Match(GetSenderName(), @"(?<=.*(?=[a-z]|[A-Z])).*").ToString();
 
             return new Payment()
             {
                 Amount = amount,
                 PhoneNumber = _phoneNumber,
-                FromAgent = fromAgent,
+                AgentName = agentName,
                 Reference = reference,
-                BankName = IPaymentService.GetBankNameFromString(BankName),
+                SenderName = SenderName,
+                BankName = IPaymentService.GetBankNameFromString(GetSenderName()),
                 ProviderName = Provider.Mpamba
             };
         }
 
         public bool HasInvalidReference()
         {
-            throw new System.NotImplementedException();
+            return (CreateReference().Length != 10);
         }
 
         public bool IsDeposit()
         {
-            throw new System.NotImplementedException();
+            return Regex.IsMatch(_message, @"((CashIn)|(Received)|(recieved)|(Deposit))");
+        }
+
+        private String CreateReference() {
```

Comment:
these are repeating in Mpamba & Airtel Money. Should they be part of the IPayment Interface?